tsconsensus: data races

https://github.com/tailscale/tailscale/actions/runs/15671654816/job/44143758116?pr=16281
```
    tsconsensus_test.go:477: waitFor success: node 0 is leader: after 4 tries
2025/06/16 04:16:53 Trying to find cluster: num targets to try: 2
2025/06/16 04:16:53 Trying to find cluster: trying 100.64.0.1
2025/06/16 04:16:53 Trying to find cluster: joined 100.64.0.1
2025-06-16T04:16:53.463Z [WARN]  raft: 1: failed to get previous log: previous-index=3 last-index=0 error="log not found"
2025/06/16 04:16:53 Trying to find cluster: num targets to try: 2
2025/06/16 04:16:53 Trying to find cluster: trying 100.64.0.2
2025-06-16T04:16:53.464Z [WARN]  raft: 0: appendEntries rejected, sending older logs: peer="{Voter 100.64.0.2 100.64.0.2:6270}" next=1
2025/06/16 04:16:53 join handler error: node is not the leader
2025/06/16 04:16:53 Trying to find cluster: could not join 100.64.0.2: remote responded 500: 
2025/06/16 04:16:53 Trying to find cluster: trying 100.64.0.1
2025-06-16T04:16:53.487Z [WARN]  raft: 2: failed to get previous log: previous-index=4 last-index=0 error="log not found"
2025/06/16 04:16:53 Trying to find cluster: joined 100.64.0.1
2025-06-16T04:16:53.488Z [WARN]  raft: 0: appendEntries rejected, sending older logs: peer="{Voter 100.64.0.3 100.64.0.3:6270}" next=1
    tsconsensus_test.go:477: waitFor success: all raft machines have all servers in their config: after 2 tries
    tsconsensus_test.go:501: waitFor success: the apply events made it into all state machines: after 2 tries
2025/06/16 04:16:56 CmdHttp stopped serving with err: http: Server closed
Error: -16T04:16:56.494Z [ERROR] raft-net: failed to decode incoming command: error="transport shutdown"
Error: -16T04:16:56.595Z [ERROR] raft-net: failed to decode incoming command: error="transport shutdown"
Error: -16T04:16:56.595Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error=EOF
Error: -16T04:16:56.663Z [ERROR] raft: 0: failed to heartbeat to: peer=100.64.0.3:6270 backoff time=10ms error="connect tcp 100.64.0.3:6270: connection was refused"
Error: -16T04:16:56.686Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error="connect tcp 100.64.0.3:6270: connection was refused"
Error: -16T04:16:56.789Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error="connect tcp 100.64.0.3:6270: connection was refused"
Error: -16T04:16:56.890Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error="connect tcp 100.64.0.3:6270: connection was refused"
Error: -16T04:16:57.001Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error="connect tcp 100.64.0.3:6270: connection was refused"
==================
WARNING: DATA RACE
Write at 0x00c00a37e028 by goroutine 4294:
  tailscale.com/tsconsensus.(*authorization).Refresh()
      /home/runner/work/tailscale/tailscale/tsconsensus/authorization.go:85 +0x23d
  tailscale.com/tsconsensus.StreamLayer.addrAuthorized()
      /home/runner/work/tailscale/tailscale/tsconsensus/tsconsensus.go:125 +0x99
  tailscale.com/tsconsensus.StreamLayer.Dial()
      /home/runner/work/tailscale/tailscale/tsconsensus/tsconsensus.go:109 +0x157
  tailscale.com/tsconsensus.(*StreamLayer).Dial()
      <autogenerated>:1 +0xd2
  github.com/hashicorp/raft.(*NetworkTransport).getConn()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:410 +0xe6
  github.com/hashicorp/raft.(*NetworkTransport).getConnFromAddressProvider()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:385 +0x64
  github.com/hashicorp/raft.(*NetworkTransport).genericRPC()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:485 +0x88
  github.com/hashicorp/raft.(*NetworkTransport).AppendEntries()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:469 +0x9e
  github.com/hashicorp/raft.(*Raft).replicateTo()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/replication.go:231 +0x35d
  github.com/hashicorp/raft.(*Raft).replicate()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/replication.go:171 +0x990
  github.com/hashicorp/raft.decodeResponse()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:774 +0x9e
  github.com/hashicorp/raft.(*NetworkTransport).genericRPC()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:501 +0x186
  github.com/hashicorp/raft.(*NetworkTransport).AppendEntries()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/net_transport.go:469 +0x9e

Goroutine 4295 (running) created at:
  github.com/hashicorp/raft.(*raftState).goFunc()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/state.go:147 +0xe4
  github.com/hashicorp/raft.(*Raft).replicate()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/replication.go:141 +0x1ca
  github.com/hashicorp/raft.(*Raft).startStopReplication.func1()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/raft.go:612 +0x38
  github.com/hashicorp/raft.(*raftState).goFunc.func1()
      /home/runner/.cache/go-mod/github.com/hashicorp/raft@v1.7.2/state.go:149 +0x8d
==================
Error: -16T04:16:57.164Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error="connect tcp 100.64.0.3:6270: connection was refused"
Error: -16T04:16:57.165Z [ERROR] raft: 0: failed to heartbeat to: peer=100.64.0.3:6270 backoff time=10ms error="connect tcp 100.64.0.3:6270: connection was refused"
Error: -16T04:16:57.397Z [ERROR] raft: 0: failed to appendEntries to: peer="{Voter 100.64.0.3 100.64.0.3:6270}" error="connect tcp 100.64.0.3:6270: connection was refused"
    tsconsensus_test.go:523: waitFor success: the apply events made it into eligible state machines: after 2 tries
2025/06/16 04:16:57 Trying to find cluster: num targets to try: 2
2025/06/16 04:16:57 Trying to find cluster: trying 100.64.0.2
2025/06/16 04:16:57 join handler error: node is not the leader
2025/06/16 04:16:57 Trying to find cluster: could not join 100.64.0.2: remote responded 500: 
2025/06/16 04:16:57 Trying to find cluster: trying 100.64.0.1
2025/06/16 04:16:57 Trying to find cluster: joined 100.64.0.1
2025-06-16T04:16:57.805Z [WARN]  raft: 2 after restarting: failed to get previous log: previous-index=6 last-index=0 error="log not found"
2025-06-16T04:16:57.806Z [WARN]  raft: 0: appendEntries rejected, sending older logs: peer="{Voter 100.64.0.3 100.64.0.3:6270}" next=1
    tsconsensus_test.go:536: waitFor success: the apply events made it into the third node when it appeared with an empty state machine: after 2 tries
2025/06/16 04:16:59 CmdHttp stopped serving with err: http: Server closed
2025/06/16 04:16:59 CmdHttp stopped serving with err: http: Server closed
2025/06/16 04:16:59 CmdHttp stopped serving with err: http: Server closed
Error: -16T04:16:59.543Z [ERROR] raft-net: failed to decode incoming command: error="read tcp 100.64.0.3:6270: operation aborted"
Error: -16T04:16:59.543Z [ERROR] raft-net: failed to decode incoming command: error="read tcp 100.64.0.3:6270: operation aborted"
Error: -16T04:16:59.574Z [ERROR] raft-net: failed to decode incoming command: error="read tcp 100.64.0.2:6270: operation aborted"
    stuntest.go:65: STUN server shutdown
    testing.go:1490: race detected during execution of test
--- FAIL: TestFollowerFailover (15.13s)
```

--- a/tsconsensus/authorization.go
+++ b/tsconsensus/authorization.go
@@ -87,29 +87,29 @@ func (a *authorization) Refresh(ctx context.Context) error {
 }
 
 func (a *authorization) AllowsHost(addr netip.Addr) bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 	if a.peers == nil {
 		return false
 	}
-	a.mu.Lock()
-	defer a.mu.Unlock()
 	return a.peers.addrs.Contains(addr)
 }
 
 func (a *authorization) SelfAllowed() bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 	if a.peers == nil {
 		return false
 	}
-	a.mu.Lock()
-	defer a.mu.Unlock()
 	return a.peers.status.Self.Tags != nil && views.SliceContains(*a.peers.status.Self.Tags, a.tag)
 }
 
 func (a *authorization) AllowedPeers() views.Slice[*ipnstate.PeerStatus] {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 	if a.peers == nil {
 		return views.Slice[*ipnstate.PeerStatus]{}
 	}
-	a.mu.Lock()
-	defer a.mu.Unlock()
 	return views.SliceOf(a.peers.statuses)
 }
 

